pkg/storage/repo/storage.go

```diff
@@ -0,0 +1,269 @@
+		fmt.Println("could not run command: ", err)
+	}
+	// otherwise, print the output from running the command
+	fmt.Println("Output: ", string(out))
+	if err != nil {
+		panic(err)
+	}
+
+	if err != nil {
+		return storage.StorageVolume{}, err
+	}
+	// }
+	filepath, err2 := url.Parse(repoURL)
+	if err2 != nil {
+		return storage.StorageVolume{}, err
+	}
+	filename := strings.Split(filepath.Path, ".")[0]
+	targetPath := "/inputs" + filename
+
+	CIDSpec, err := sp.Upload(ctx, outputPath)
+	// If estuary key exists then upload it to estuary
+	if err != nil {
+		return storage.StorageVolume{}, err
+	}
+	CID := CIDSpec.CID
+	// Update the KV store
+	SHA1HASH, _ := UrltoLatestCommitHash(repoURL)
+	spkey := sp.EstuaryAPIKey
+	envkey := os.Getenv("ESTUARY_API_KEY")
+	var key string
+	if spkey != "" && envkey == "" {
+		key = spkey
+	} else if envkey != "" && spkey == "" {
+		key = envkey
+	} else if envkey != "" && spkey != "" {
+		key = spkey
+	} else {
+		key = ""
+	}
+	if key != "" {
+		log1.Println("Pinning to Estuary...")
+		//nolint:govet,noctx
+		err := estuary.PinToIPFSViaEstuary(ctx, key, CID)
+		if err != nil {
+			return storage.StorageVolume{}, err
+		}
+		log1.Println("Successfully Pinned to Estuary...")
+	}
+	data := url.Values{}
+	data.Set("key", SHA1HASH)
+	data.Set("value", CID)
+	if err != nil {
+		return storage.StorageVolume{}, err
+	}
+	err = CreateSHA1CIDPair(data)
+	if err != nil {
+		fmt.Printf("Failed to create SHA1CIDPair: %v", err)
+	}
+	// return the volume
+	volume := storage.StorageVolume{
+		Type:   storage.StorageVolumeConnectorBind,
+		Source: outputPath,
+		Target: targetPath,
+	}
+
+	return volume, nil
+}
+
+func (sp *StorageProvider) Upload(ctx context.Context, localPath string) (model.StorageSpec, error) {
+	ctx, span := system.GetTracer().Start(ctx, "storage/repo/apicopy.Upload")
+	defer span.End()
+
+	cid, err := sp.IPFSClient.Upload(ctx, localPath)
+	fmt.Print(cid)
+	if err != nil {
+		fmt.Print(err)
+		return model.StorageSpec{}, err
+	}
+	return model.StorageSpec{
+		StorageSource: model.StorageSourceIPFS,
+		CID:           cid.CID,
+	}, nil
+}
+
+func (sp *StorageProvider) CleanupStorage(
+	ctx context.Context,
+	_ model.StorageSpec,
+	volume storage.StorageVolume,
+) error {
+	_, span := system.GetTracer().Start(ctx, "pkg/storage/repo/repo.CleanupStorage")
+	defer span.End()
+
+	pathToCleanup := filepath.Dir(volume.Source)
+	log.Ctx(ctx).Debug().Str("Path", pathToCleanup).Msg("Cleaning up")
+	return os.RemoveAll(pathToCleanup)
+}
+
+func (sp *StorageProvider) Explode(_ context.Context, spec model.StorageSpec) ([]model.StorageSpec, error) {
+	return []model.StorageSpec{
+		{
+			Name:          spec.Name,
+			StorageSource: model.StorageSourceRepoClone,
+			Path:          spec.Path,
+			Repo:          spec.Repo,
+		},
+	}, nil
+}
+
+func CreateSHA1CIDPair(data url.Values) error {
+	//nolint:gomnd
+	time.Sleep(100 * time.Millisecond)
+	//nolint:noctx
+	resp, err := http.PostForm("http://kv.bacalhau.org", data)
```

Comment:
What is this service and where does the code for it live?

Reply:
https://github.com/bacalhau-project/kv.bacalhau.org